File: BrainMie/src/BrainMie.Core/Pipeline.cs
namespace BrainMie.Core;

using BrainMie.Core.Data;

/// <summary>
/// End-to-end pipeline: .dmt file → corrected isotopic peak list.
/// </summary>
public static class Pipeline
{
    /// <summary>
    /// Read a .dmt SQLite file, convert all ions to neutral mass, detect MIE artifacts
    /// across all charge states in one unified mass spectrum, and return a flat list of
    /// processing rows containing observed and suppression-corrected peaks.
    /// </summary>
    /// <param name="filePath">Path to the .dmt file.</param>
    /// <param name="binWidth">Mass bin width in Da for ion-event aggregation. Default 0.02 Da.</param>
    /// <param name="maxGap">Max consecutive missing isotope peaks per cluster. Default 3.</param>
    /// <param name="massMatchTol">
    /// Absolute mass tolerance in Da for matching peaks to isotope positions (replaces
    /// the former per-charge ppm parameter). Default 0.2 Da.
    /// </param>
    /// <param name="minClusterPeaks">Minimum observed peaks to retain any cluster. Default 8.</param>
    /// <param name="nPeaks">Number of theoretical isotope peaks to generate. Default 25.</param>
    /// <param name="suppressionThreshold">
    /// Fraction of fitted theoretical intensity below which an observed peak is
    /// considered MIE-suppressed. Default 0.5 (50%).
    /// </param>
    /// <param name="minFitR2">
    /// Minimum R² of the single-species averagine fit required to keep a cluster.
    /// Default 0.5.
    /// </param>
    /// <returns>
    /// A tuple of the corrected row list and the original raw ion list (needed by
    /// <see cref="ProteoformAggregator"/> to derive charge-state annotations).
    /// </returns>
    public static (List<ProcessingRow> Rows, List<(double Mz, int Charge, double? Intensity)> Ions)
        ProcessDmt(
            string filePath,
            double binWidth             = 0.02,
            int    maxGap               = 3,
            double massMatchTol         = 0.2,
            int    minClusterPeaks      = 8,
            int    nPeaks               = 25,
            double suppressionThreshold = 0.5,
            double minFitR2             = 0.5)
    {
        var ions = DmtReader.ReadIons(filePath);

        // Convert every ion to neutral mass immediately.
        // mass = mz × charge − charge × HMass
        List<(double Mass, double Intensity)> peaks;
        if (ions.Count > 0 && ions[0].Intensity.HasValue)
        {
            // File already stores per-peak intensities — use them directly.
            peaks = ions
                .Select(i => (Mass: i.Mz * i.Charge - i.Charge * Constants.HMass, Intensity: i.Intensity!.Value))
                .OrderBy(p => p.Mass)
                .ToList();
            peaks = EnvelopeDetector.PickLocalMaxima(peaks);
        }
        else
        {
            var massValues = ions.Select(i => i.Mz * i.Charge - i.Charge * Constants.HMass);
            peaks = EnvelopeDetector.BinIons(massValues, binWidth);
            peaks = EnvelopeDetector.PickLocalMaxima(peaks);
        }

        var clusters = EnvelopeDetector.FindClusters(peaks, maxGap, massMatchTol);

        var results = new List<ProcessingRow>();

        foreach (var cluster in clusters)
        {
            // Minimum isotope count
            if (cluster.Peaks.Count < minClusterPeaks) continue;

            // Classify first — needed for both R² filter and theory-based apex index.
            var classification = Classifier.Classify(cluster, nPeaks);
            if (classification.FitRSquared < minFitR2) continue;

            // Use the theoretical averagine apex (adjusted by fit offset) for the
            // completeness check. This is robust to MIE suppression of the centroid
            // peak, which would shift the observed intensity maximum to a flanking peak.
            var theoretical  = IsotopeDistribution.ComputeMassEnvelope(cluster.NeutralMass, nPeaks);
            if (theoretical.Count == 0) continue;
            int theoApexTi   = theoretical.Select((t, i) => (t.Intensity, i)).MaxBy(x => x.Intensity)!.i;
            int apexIdx      = theoApexTi - classification.FitOffset;
            var obsSet       = cluster.Peaks.Select(p => p.IsotopeIndex).ToHashSet();
            bool gapNearApex = Enumerable.Range(apexIdx - 2, 5)
                                   .Where(i => i >= 0)
                                   .Any(i => !obsSet.Contains(i));
            if (gapNearApex) continue;

            var suppressed    = Classifier.DetectSuppressed(cluster, suppressionThreshold, nPeaks);
            var suppressedMap = suppressed.ToDictionary(s => s.IsotopeIndex, s => s.CorrectedIntensity);

            double clusterIonCount   = cluster.Peaks.Sum(p => p.Intensity);
            double suppressionDelta  = suppressed.Sum(s => s.CorrectedIntensity - s.ObservedIntensity);
            double correctedIonCount = clusterIonCount + suppressionDelta;

            foreach (var peak in cluster.Peaks)
            {
                bool    isSuppressed       = suppressedMap.TryGetValue(peak.IsotopeIndex, out double ci);
                double? correctedIntensity = isSuppressed ? ci : null;

                results.Add(new ProcessingRow(
                    IsotopeMass:        peak.Mass,
                    Intensity:          peak.Intensity,
                    NeutralMass:        cluster.NeutralMass,
                    IsotopicIndex:      peak.IsotopeIndex,
                    IsEstimated:        false,
                    IsSuppressed:       isSuppressed,
                    EstimatedIntensity: null,
                    CorrectedIntensity: correctedIntensity,
                    Uncertainty:        null,
                    Confidence:         classification.Confidence,
                    Hypothesis:         classification.Hypothesis,
                    FitRSquared:        classification.FitRSquared,
                    GapAtApex:          classification.GapAtApex,
                    Notes:              classification.Notes,
                    ClusterIonCount:    clusterIonCount,
                    CorrectedIonCount:  correctedIonCount));
            }
        }

        return (results, ions);
    }
}

File: brainpy/mie/pipeline.py
"""
End-to-end pipeline: .dmt file → corrected isotopic peak table.

Typical usage
-------------
    from brainpy.mie import process_dmt

    df = process_dmt("experiment.dmt")
    # All observed peaks plus reconstructed MIE peaks in one DataFrame.

    # Only the reconstructed (rescued) peaks:
    rescued = df[df["IsEstimated"]]

    # Write corrected data back to SQLite:
    import sqlite3
    conn = sqlite3.connect("corrected.dmt")
    df.to_sql("CorrectedIon", conn, if_exists="replace", index=False)
    conn.close()
"""
import pandas as pd
from typing import Tuple

from .reader import read_ions
from .envelope import bin_ions, find_isotopic_clusters
from .classifier import classify
from .reconstructor import reconstruct


def process_dmt(
    filepath: str,
    charge_range: Tuple[int, int] = (1, 10),
    bin_width: float = 0.005,
    max_gap: int = 3,
    ppm_tolerance: float = 20.0,
    min_cluster_peaks: int = 3,
    npeaks: int = 25,
) -> pd.DataFrame:
    """
    Read a .dmt SQLite file and return a corrected isotopic peak DataFrame.

    Parameters
    ----------
    filepath : str
        Path to the .dmt file.
    charge_range : (int, int)
        Inclusive range of charge states to process.
    bin_width : float
        m/z bin width in Da used to aggregate individual ion events into peaks.
        Should be << 1/max_charge.  Default 0.005 Da.
    max_gap : int
        Maximum number of consecutive missing isotope peaks that are still
        treated as belonging to the same cluster.  Default 3.
    ppm_tolerance : float
        m/z tolerance in ppm for assigning peaks to isotope positions.
    min_cluster_peaks : int
        Clusters with fewer observed peaks than this (and no gaps) are skipped.
    npeaks : int
        Number of theoretical isotope peaks to generate per cluster.

    Returns
    -------
    pd.DataFrame with columns:

    Mz                  — observed m/z (NaN for estimated peaks)
    Charge              — charge state
    Intensity           — observed ion count (NaN for estimated peaks)
    NeutralMass         — estimated neutral mass of the cluster
    IsotopicIndex       — 0-based index within the isotopic series
    IsEstimated         — True for reconstructed MIE peaks
    EstimatedIntensity  — predicted ion count (populated for estimated peaks)
    Uncertainty         — 1-sigma prediction interval of EstimatedIntensity
    Confidence          — 0–1 score for the stated hypothesis / reconstruction
    Hypothesis          — 'mie', 'overlap', 'ambiguous', or 'no_gap'
    FitRSquared         — R² of the single-species averagine fit
    GapAtApex           — whether the gap sits at the theoretical envelope apex
    Notes               — human-readable caveats
    """
    ions_df = read_ions(filepath)
    records = []

    for charge in range(charge_range[0], charge_range[1] + 1):
        subset = ions_df[ions_df["Charge"] == charge]
        if subset.empty:
            continue

        if "Intensity" in subset.columns:
            # Pre-aggregated peaks: use Intensity directly
            peaks = list(zip(subset["Mz"].values, subset["Intensity"].values))
        else:
            # Individual ion events: bin by m/z to derive peak intensities
            peaks = bin_ions(subset["Mz"].values, bin_width=bin_width)

        clusters = find_isotopic_clusters(
            peaks,
            charge,
            max_gap=max_gap,
            ppm_tolerance=ppm_tolerance,
        )

        for cluster in clusters:
            n_obs = len(cluster.peaks)
            if n_obs < min_cluster_peaks and not cluster.gap_indices:
                continue

            classification = classify(cluster, npeaks=npeaks)
            recon_peaks = reconstruct(cluster, classification, npeaks=npeaks)

            # --- Observed peaks -------------------------------------------
            for peak in cluster.peaks:
                records.append(
                    {
                        "Mz": peak.mz,
                        "Charge": peak.charge,
                        "Intensity": peak.intensity,
                        "NeutralMass": cluster.neutral_mass,
                        "IsotopicIndex": peak.isotope_index,
                        "IsEstimated": False,
                        "EstimatedIntensity": None,
                        "Uncertainty": None,
                        "Confidence": classification.confidence,
                        "Hypothesis": classification.hypothesis,
                        "FitRSquared": classification.fit_r_squared,
                        "GapAtApex": classification.gap_at_apex,
                        "Notes": classification.notes,
                    }
                )

            # --- Reconstructed (rescued) peaks ----------------------------
            for rec in recon_peaks:
                records.append(
                    {
                        "Mz": None,
                        "Charge": rec.charge,
                        "Intensity": None,
                        "NeutralMass": cluster.neutral_mass,
                        "IsotopicIndex": rec.isotope_index,
                        "IsEstimated": True,
                        "EstimatedIntensity": rec.estimated_intensity,
                        "Uncertainty": rec.uncertainty if rec.uncertainty != float("inf") else None,
                        "Confidence": rec.confidence,
                        "Hypothesis": classification.hypothesis,
                        "FitRSquared": classification.fit_r_squared,
                        "GapAtApex": classification.gap_at_apex,
                        "Notes": classification.notes,
                    }
                )

    return pd.DataFrame(records)
